rust/feature-flags/src/cohort_operations.rs

```diff
@@ -0,0 +1,369 @@
+            // but I'm not sure if accessing Cohort definitions should be a FlagError (vs idk, a CohortError?  A more general API error?)
+            FlagError::DatabaseUnavailable
+        })?;
+
+        let query = "SELECT id, name, description, team_id, deleted, filters, query, version, pending_version, count, is_calculating, is_static, errors_calculating, groups, created_by_id FROM posthog_cohort WHERE id = $1 AND team_id = $2";
+        let cohort = sqlx::query_as::<_, Cohort>(query)
+            .bind(cohort_id)
+            .bind(team_id)
+            .fetch_optional(&mut *conn)
+            .await
+            .map_err(|e| {
+                tracing::error!("Failed to fetch cohort from database: {}", e);
+                FlagError::Internal(format!("Database query error: {}", e))
+            })?;
+
+        cohort.ok_or_else(|| {
+            FlagError::CohortNotFound(format!(
+                "Cohort with id {} not found for team {}",
+                cohort_id, team_id
+            ))
+        })
+    }
+
+    /// Returns all cohorts for a given team
+    #[instrument(skip_all)]
+    pub async fn list_from_pg(
+        client: Arc<dyn DatabaseClient + Send + Sync>,
+        team_id: i32,
+    ) -> Result<Vec<Cohort>, FlagError> {
+        let mut conn = client.get_connection().await.map_err(|e| {
+            tracing::error!("Failed to get database connection: {}", e);
+            FlagError::DatabaseUnavailable
+        })?;
+
+        let query = "SELECT id, name, description, team_id, deleted, filters, query, version, pending_version, count, is_calculating, is_static, errors_calculating, groups, created_by_id FROM posthog_cohort WHERE team_id = $1";
+        let cohorts = sqlx::query_as::<_, Cohort>(query)
+            .bind(team_id)
+            .fetch_all(&mut *conn)
+            .await
+            .map_err(|e| {
+                tracing::error!("Failed to fetch cohorts from database: {}", e);
+                FlagError::Internal(format!("Database query error: {}", e))
+            })?;
+
+        Ok(cohorts)
+    }
+
+    /// Parses the filters JSON into a CohortProperty structure
+    // TODO: this doesn't handle the deprecated "groups" field, see
+    // https://github.com/PostHog/posthog/blob/feat/dynamic-cohorts-rust/posthog/models/cohort/cohort.py#L114-L169
+    // I'll handle that in a separate PR.
+    pub fn parse_filters(&self) -> Result<Vec<PropertyFilter>, FlagError> {
+        let cohort_property: CohortProperty = serde_json::from_value(self.filters.clone())
+            .map_err(|e| {
+                tracing::error!("Failed to parse filters for cohort {}: {}", self.id, e);
+                FlagError::CohortFiltersParsingError
+            })?;
+        Ok(cohort_property
+            .properties
+            .to_property_filters()
+            .into_iter()
+            .filter(|f| !(f.key == "id" && f.prop_type == "cohort"))
```

Comment:
You can use `retain` here, like:
```rust
        let mut props = cohort_property.properties.to_property_filters();
        props.retain(|f| !(f.key == "id" && f.prop_type == "cohort"));
        Ok(props)
```

Or you can make `to_property_filters` return an `impl Iter<Item = PropertyFilter>` and then do the `filter().collect()` as you already do - collecting to a vec just to `into_iter` and then collect again is an antipattern.